### DISCUSSION TITLE
GitHub Apps "Client secrets" vs "Private keys"

### DISCUSSION THREAD

**Discussion by ChrisAcrobat** [community/community#28581] [New to GitHub] [answered]
What is the difference between `Client secrets` and `Private keys` in GitHub Apps? Why and when should I chose the one over the other?
Thanks for the help!

**Comment by ihnatenkoo** [comment]
Hello Christoffer, maybe this article can help you:
**https://blog.axway.com/learning-center/digital-security/keys-oauth/api-keys-oauth#:~:text=The%20API%20key%20ID%20is,languages%20and%20frameworks%20provide%20support.**

**Reply by ChrisAcrobat** [reply]
Sorry, but no, I still don't understand the difference. Both `Client secrets` and `Private keys` are only known to the client and the API, are they not? To me they are just two different ways of resolving a access token, but I feel that it should be more too it than just that.

**Comment by jamesblashill** [comment]
GitHub Apps use cryptographic keys for secure communication between the app and GitHub. The client secret and private key are two types of keys used in GitHub Apps, and they serve different purposes.

A client secret is a shared secret between the GitHub App and GitHub. It's used to authenticate the app when it interacts with GitHub's API. The client secret is generated when the app is installed on a GitHub account and can be found in the app's settings. The client secret should be kept confidential and not be shared with anyone other than authorized users of the app.

A private key, on the other hand, is used for signing and encrypting data. It's unique to each installation of a GitHub App and is used to verify that requests coming from GitHub are from the app that claims to be making them. Private keys are generated when a GitHub App is created and are stored securely on the app's server. Private keys should also be kept confidential and not be shared with anyone other than authorized users of the app.

In summary, the client secret is used for authentication, while the private key is used for verification and encryption. Both are essential components of a GitHub App's security infrastructure and should be kept confidential and secure.

**Reply by Eeebru** [reply]
GPT bullock response??

**Reply by ChrisAcrobat** [reply]
@Eeebru Is it wrong?

**Reply by cpplain** [reply]
The answer provided by @Prid13 is the correct answer.

**Reply by ChrisAcrobat** [reply]
@cpplain Okay, I'll change. But for future reference, please explain why.

**Comment by Prid13** [comment, marked as answer]
* **Private Keys** are used to generate _**installation**_ access tokens that log all activities performed with that token _**as the Github App itself**_ (the activities will be attributed to the App)
  * e.g., an issue/comment posted via an access token generated using a Private Key, will show up as a "Bot" comment, with the bot's username being the App's name
* **Client Secrets** are used to generate _**user**_ access tokens that log all activities performed with that token _**as the user themselves**_ (the activities will be attributed to the user, as if the user made it themselves)
  * e.g., an issue/comment posted via a _user access token_ generated with a Client Secret, will show up as having been posted by the user themselves

> _**Source:** https://docs.github.com/en/enterprise-cloud@latest/apps/creating-github-apps/about-creating-github-apps/best-practices-for-creating-a-github-app#installation-access-tokens-user-access-tokens-and-refresh-tokens_

> Installation access tokens are used to make API requests on behalf of an app installation. User access tokens are used to make API requests on behalf of a user. Refresh tokens are used to regenerate user access tokens. **Your app can use its private key to generate an installation access token. Your app can use its client secret to generate a user access token and refresh token.**

**Reply by Prid13** [reply]
_Example: issue posted using an installation access token (generated with a Private Key):_
> <img src="https://github.com/community/community/assets/27897747/d2eb1da8-0b23-400f-912d-4e8695e2789d" width="400">
